builtin/providers/aws/auth_helpers_test.go

```diff
@@ -586,9 +587,9 @@ func invalidAwsEnv(t *testing.T) func() {
 	return ts.Close
 }
 
-// getMockedAwsIamApi establishes a httptest server to simulate behaviour
-// of a real AWS' IAM server
-func getMockedAwsIamApi(endpoints []*iamEndpoint) (func(), *iam.IAM) {
+// getMockedAwsIamStsApi establishes a httptest server to simulate behaviour
+// of a real AWS' IAM & STS server
+func getMockedAwsIamStsApi(endpoints []*iamEndpoint) (func(), *iam.IAM, *sts.STS) {
```

Comment:
Would you mind also adding the actual STS mock responses (for success + error) and two tests testing STS?

1. `iam:GetUser` - 403 (failure) `=>` `sts:GetCallerIdentity` - 200 (success)
2. `iam:GetUser` - 403 (failure) `=>` `sts:GetCallerIdentity` - [any failure](http://docs.aws.amazon.com/STS/latest/APIReference/CommonErrors.html) `=>` `iam:ListRoles` - 200 (successful fallback to ListRoles)

You can use examples of responses from the API reference as mocks: http://docs.aws.amazon.com/STS/latest/APIReference/API_GetCallerIdentity.html